### PR TITLE
Add scalar python types to array-like types

### PR DESCRIPTION
**Pull request by romanngg**
This will make `isinstance` called on scalars return True for both jitted and non-jitted JAX programs. Note that
`isinstance(x, onp.ndarray)` is False for these types, so it breaks this correspondence.

#1080

### CODE REVIEW

**Review by mattjj — APPROVED** [review]
LGTM!

You could check `six.PY3` and include `long` if we're in PY2. WDYT?

File: jax/numpy/lax_numpy.py [hunk 1]
```diff
@@ -80,7 +80,12 @@ def removechars(s, chars):
 # array base class). We can override the isinstance behavior directly, without
 # having the complexity of multiple inheritance on those classes, by defining
 # the ndarray class to have a metaclass with special __instancecheck__ behavior.
-_arraylike_types = (onp.ndarray, UnshapedArray, DeviceArray)
+#
+# Note that scalar python types are included to have consistent isinstance
+# behavior when using JIT. This is different from numpy which treats scalars
+# differently (i.e. isintance(1, onp.ndarray) == False).
```
Comment (jekbradbury) [inline jax/numpy/lax_numpy.py]:
```suggestion
# differently (i.e. isinstance(1, onp.ndarray) == False).
```

**Comment by romanngg** [comment]
Thanks Matt and James, added your suggestions!

**Comment by shoyer** [comment]
It's a little surprising that types change under `jit`, but it sort of makes sense if you think about it -- and I fear the fix here might be worse. What if I'm using `isinstance` checks to see if I can safely use NumPy attributes, like `dtype` and `shape`?

**Comment by mattjj** [comment]
@shoyer let's keep discussing this! Though we should keep in mind we're going for "best reasonable solution" rather than "perfect."

Attributes like `dtype` and `shape` work under a `jit` (because scalars are always 0d arrays) but not without a `jit`. But that doesn't seem so bad because NumPy ascribes dtypes and shapes to Python scalars already (i.e. the output of `onp.result_type` and `onp.shape` respectively). A separate argument is that using `isinstance` to look for attributes arguably goes against the Python duck-typing spirit.

But I am concerned if you think this 'fix' is worse. Can you say more about how this might be worse? Is it about diverging further from NumPy/Python scalar behavior, or rather about having different behavior under `jit` (specifically around isinstance and how it relates to attributes being available)?

Maybe another principle we should include here is the principle of least surprise. Is `isinstance(3, np.ndarray)` surprising?

**Comment by shoyer** [comment]
Honestly, I'm a little amazed that JAX manages to make `isinstance` with `jax.numpy.ndarray` work at all :). 

Perhaps we should think about use cases? For the use case of checking for arrays/scalars vs. other non-primitive Python things inside some sort of nested Python data structure, `isinstance(3, np.ndarray)` seems perfectly reasonable. So maybe this is fine after all...

**Comment by shoyer** [comment]
I cannot think of circumstances when it would be useful to distinguish between Python scalars and arrays with JAX, with the possible exception of value-based casting (and that can be handled by some separate mechanism). It does seems like something like `np.is_number_or_array()` would be the right way to handle this, but NumPy doesn't provide that API 🤷‍♂

**Comment by hawkinsp** [comment]
Closing this PR, since it's fairly stale at this point.

I think #1080 is to some extent working as intended; `jit` has the effect of promoting scalars to `ndarrays`. So far we are mostly happy with that choice, but it is not completely transparent to introspection tools like `isinstance`.

**Pull request closed without merging**